README.md
<h1 align="center" style="border-bottom: none;">:ok_person::crayon::speech_balloon:&nbsp;&nbsp; Act, Draw, Explain &nbsp;&nbsp;:speech_balloon::crayon::ok_person:</h1>
<h3 align="center">A multi-player game based on popular games, such as Charades, Heads-up and Pictionary</h3>

<p align="center">
    <img alt="License" src="https://img.shields.io/github/license/radeklat/act-draw-explain">
    <img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/radeklat/act-draw-explain">
    <img alt="App version" src="https://img.shields.io/badge/dynamic/yaml?color=informational&label=version&query=version&url=https%3A%2F%2Fraw.githubusercontent.com%2Fradeklat%2Fact-draw-explain%2Fmaster%2Fpubspec.yaml">
    <a href="https://codecov.io/gh/radeklat/act-draw-explain">
        <img alt="codecov" src="https://codecov.io/gh/radeklat/act-draw-explain/branch/master/graph/badge.svg">
    </a>
    <a href="https://weblate.lat.sk/engage/act-draw-explain/?utm_source=github">
        <img alt="Translation status" src="https://weblate.lat.sk/widgets/act-draw-explain/-/svg-badge.svg">
    </a>
</p>

# How to contribute

This is an open source project and everyone is welcome to contribute to the game. Here are several options how you can contribute today:

1. [Report bugs](https://github.com/radeklat/act-draw-explain/issues/new/choose)
2. [Suggest new features](https://github.com/radeklat/act-draw-explain/issues/new/choose)
3. Create new topics, add more questions into existing topics or fix typos (see [Game data](#game-data) below)
4. Translate the app into more languages (see [Guide for translators](#guide-for-translators) below)
5. Improve the code base via Pull requests

You don't have to be a Flutter developer to contribute. Files are editable in GitHub directly and translations are handled by a translation system. Easier task, such as fixing a typo, are a matter of a few click.

# Guide for translators

1. Register to [Weblate](https://weblate.lat.sk/engage/act-draw-explain/) for the Act, Draw, Explain app.
2. Translate one of the following components:
   * [User interface](https://weblate.lat.sk/projects/act-draw-explain/flutter-app/) (menus, help, etc.)
   * [Game topics](https://weblate.lat.sk/projects/act-draw-explain/game-topics/)
   * [Game questions](https://weblate.lat.sk/projects/act-draw-explain/game-questions/)

Overall language support:

[![Translation status](https://weblate.lat.sk/widgets/act-draw-explain/-/multi-auto.svg)](https://weblate.lat.sk/engage/act-draw-explain/?utm_source=widget)

## Disabling translations

Not all questions or topics make sense to translate. Some may be culturally specific or have no equivalent in the target language. To disable a question or a topic, use `DISABLED` as the translation string. It will be then hidden by the app.

## Adding new language

1. Start new translations in [Weblate](https://weblate.lat.sk/projects/act-draw-explain/) for all components.
2. Translate strings in all components.

## Plurals and genders

Some strings may allow specifying variants for different plurals or genders. Unfortunately, the translation system renders these in their original syntax. See [Plural](https://support.crowdin.com/icu-message-syntax/#plural) and [gender](https://support.crowdin.com/icu-message-syntax/#select) for a reference.

# Development

This project is written in Flutter and at the moment being built and distributed only on Android. However, the language was chosen with long term intention to distribute the project to other platforms as well.

The development language is English to make this project as widely available as possible to the development community. This includes commit messages, change logs, development documentation and code comments.

## Translation support in code

The application utilizes [`intl`](https://pub.dev/packages/intl) and [`intl_utils`](https://pub.dev/packages/intl_utils) packages.

Application translations are kept in [ARB files](https://github.com/google/app-resource-bundle/wiki/ApplicationResourceBundleSpecification). See also [Plurals and genders](#plurals-and-genders).

Locale is in form of `<LANGUAGE CODE>` or `<LANGUAGE CODE>_<COUNTRY CODE>`, where:
* `<LANGUAGE CODE>` is [ISO 639-1](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes) language code
* `<COUNTRY CODE>` is [ISO 3166-1 alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2#Officially_assigned_code_elements) country code. This code is optional.

### Adding new UI strings

1. Add the string into `lib/I10n/intl_en.arb` (this is the base language). See also  [Plurals and genders](#plurals-and-genders).
2. Run `flutter pub run intl_utils:generate` to generate boilerplate code
3. Use the boilerplate in the codebase as `S.of(context).NAME` where `NAME` is name of the string.
4. Translate the string to other languages.

## Game data

Topic and questions are at the moment distributed with the app. They are located in `assets/data`. The `topics.xliff` and `questions.xliff` are reference files that must contain all available strings in any language (`en`, if possible). They contain additional metadata used only by the app and ignored by the translation system.

Running `flutter pub run sync_translation_files` from the repository source will:
* inject `DISABLED` into all question of a disabled topic in given language variant (see [Disabling translations](#disabling-translations) below)
* inject topic names into questions to be visible as context in the translation system

## Testing

Since testing tools for Flutter are still in their infancy, the tooling is often broken, too simple or non-existent. To see test coverage, the project utilizes codecov.io for visualisation.

Unit tests are located in `test/unit`, widget tests in `test/widget`. To generate a combined test/widget tests coverage report, run:

```bash
flutter test --coverage && bash <(curl -s https://codecov.io/bash) -f coverage/lcov.info
```



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
Types of changes are:

* **Breaking changes** for breaking changes.
* **Features** for new features or changes in existing functionality.
* **Fixes** for any bug fixes.

## [Unreleased]
<!-- Don't forget to update version in pubspec.yaml -->

## [2.2.0+20] - 2020-11-24 <a name="2.2.0" />
### Features
* Application font replaced with a Google font Ubuntu
* New license page in the About dialog
* New act, draw, explain game mode

## [2.1.0+18] - 2020-07-19 <a name="2.1.0" />
### Features
* New topics:
  * Czech
    * National dishes
  * English
    * Countries of Europe
    * Board Games
    * Visual Styles
    * Sports
    * Animals
    * Hobbies
    * Food and drinks
    * Occupations
    * Musical instruments

### Fixes
* Remove duplicated questions
* Fixed typos

## [2.0.0+17] - 2020-07-19 <a name="2.0.0" />
### Features
* Translate application to English
* Allow choosing language for the app UI and game data
* Translate some topics into English
* App UI and Game data can be translated in a [translation system](https://weblate.lat.sk)
* Custom crash screen

### BREAKING CHANGES
* Topics that have no translations in selected language are hidden
* Topics disabled for selected language are hidden

## [1.1.1+15] - 2020-07-11 <a name="1.1.1" />
### Fixes
* Crash when system language was not available among translations. Now the app UI will default to English.

## [1.1.0+13] - 2020-06-20 <a name="1.1.0" />
### Features
* Adds UI translation to English

### Fixes
* Typos

## [1.0.0+11] - 2020-06-20 <a name="1.0.0" />
### Features
* Adds attributions screen

### BREAKING CHANGES
* Change log link moved into "About app" screen

## [0.7.0+10] - 2020-05-31 <a name="0.7.0" />
### Features
* New topics "Occupations", "Assigned reading", "Musical instruments", "Best selling singles" and "Czech companies"

### Fixes
* Typos
* Duplicate questions removed

## [0.6.0+9] - 2020-05-24 <a name="0.6.0" />
### Features
* New topics "Videogames" and "Food and drinks"
* Adds new questions into "RuPaul's Drag Race"

## [0.5.0+8] - 2020-05-17 <a name="0.5.0" />
### Features
* New topic "hobbies"

### Fixes
* Remove unnecessary and PID information from bug report
* GameScore counting guessed questions count wrong on timeout

## [0.4.0+7] - 2020-05-13 <a name="0.4.0" />
### Features
* Link to feedback (bug report, feature requests) from the app
* Link to change log from the app

### Fixes
* Font od body text from bold to normal

## [0.3.0+6] - 2020-05-10 <a name="0.3.0" />
### Features
* Analytics

### Fixes
* Wrong score when the last question was a pass

## [0.2.1+5] - 2020-05-09 <a name="0.2.1" />
### Features
*  Downgraded Flutter from v0.17.0 causing app to crash on start to previous v1.16.3

## [0.2.0+4] - 2020-05-09 <a name="0.2.0" />
### Features
* App name localisation to CS/EN in app launcher
* Update application icon

### Fixes
* Overflowing cards with topic
* App crashing when max questions higher than available questions count.
* Missing advanced game play documentation

## [0.1.2+3] - 2020-05-08 <a name="0.1.2" />

### Features
* Basic analytics
### Fixes
* Crashing countdown screen causing it to be stuck on the initial value.

## [0.1.1+2] - 2020-05-08 <a name="0.1.1" />

### Features
* Crashlytics

## [0.1.0+1] - 2020-04-08 <a name="0.1.0" />

### Features
* Initial source code

[Unreleased]: https://github.com/radeklat/act-draw-explain/compare/releases/2.2.0+20...HEAD
[2.2.0+20]: https://github.com/radeklat/act-draw-explain/compare/releases/2.1.0+18...releases/2.2.0+20
[2.1.0+18]: https://github.com/radeklat/act-draw-explain/compare/releases/2.0.0+17...releases/2.1.0+18
[2.0.0+17]: https://github.com/radeklat/act-draw-explain/compare/releases/1.1.1+15...releases/2.0.0+17
[1.1.1+15]: https://github.com/radeklat/act-draw-explain/compare/releases/1.1.0+13...releases/1.1.1+15
[1.1.0+13]: https://github.com/radeklat/act-draw-explain/compare/releases/1.0.0+11...releases/1.1.0+13
[1.0.0+11]: https://github.com/radeklat/act-draw-explain/compare/releases/0.7.0+10...releases/1.0.0+11
[0.7.0+10]: https://github.com/radeklat/act-draw-explain/compare/releases/0.6.0+9...releases/0.7.0+10
[0.6.0+9]: https://github.com/radeklat/act-draw-explain/compare/releases/0.5.0+8...releases/0.6.0+9
[0.5.0+8]: https://github.com/radeklat/act-draw-explain/compare/releases/0.4.0+7...releases/0.5.0+8
[0.4.0+7]: https://github.com/radeklat/act-draw-explain/compare/releases/0.3.0+6...releases/0.4.0+7
[0.3.0+6]: https://github.com/radeklat/act-draw-explain/compare/releases/0.2.1+5...releases/0.3.0+6
[0.2.1+5]: https://github.com/radeklat/act-draw-explain/compare/releases/0.2.0+4...releases/0.2.1+5
[0.2.0+4]: https://github.com/radeklat/act-draw-explain/compare/releases/0.1.2+3...releases/0.2.0+4
[0.1.2+3]: https://github.com/radeklat/act-draw-explain/compare/releases/0.1.1+2...releases/0.1.2+3
[0.1.1+2]: https://github.com/radeklat/act-draw-explain/compare/releases/0.1.0+1...releases/0.1.1+2
[0.1.0+1]: https://github.com/radeklat/act-draw-explain/compare/initial...releases/0.1.0+1

lib/data/topics.dart
import 'dart:collection';

import 'package:act_draw_explain/models/topic.dart';
import 'package:collection/collection.dart';
import 'package:flutter/material.dart';

UnmodifiableSetView<int> range(int min, int max, {Set<int> extra}) {
  return UnmodifiableSetView(Set.from([for (var i = min; i <= max; i += 1) i, ...(extra ?? {})]));
}

HashMap<int, Topic> legacyTopics = HashMap.fromIterable(
  [
    Topic(
      id: 2,
      text: "Brno",
      color: Colors.red,
      icon: Icon(Icons.location_city),
      questionIDs: range(21, 49, extra: {264, 265}),
    ),
    Topic(
      id: 1,
      text: "Pohádky",
      color: Colors.lightBlueAccent,
      icon: Icon(Icons.ac_unit),
      questionIDs: range(1, 20, extra: range(355, 384)),
    ),
    Topic(
      id: 3,
      text: "Zahraniční seriály",
      color: Colors.deepPurple,
      icon: Icon(Icons.tv),
      questionIDs: range(50, 64),
    ),
    Topic(
      id: 5,
      text: "ČSFD Top 20 Filmů",
      color: Colors.redAccent,
      icon: Icon(Icons.tv),
      questionIDs: range(85, 103, extra: {76}),
    ),
    Topic(
      id: 6,
      text: "ČSFD Top 20 Seriálů",
      color: Colors.redAccent,
      icon: Icon(Icons.tv),
      questionIDs: range(105, 124),
    ),
    Topic(
      id: 7,
      text: "RuPaul's Drag Race",
      color: Colors.pinkAccent,
      icon: Icon(Icons.looks),
      questionIDs: range(125, 172, extra: {435, 436, 901, 902, 903, 1109, 1110, 1111, 1112, 1113, 1114, 1115}),
    ),
    Topic(
        id: 8,
        text: "Top Android Apps",
        color: Colors.green,
        icon: Icon(Icons.android),
        questionIDs: range(173, 201),
        sources: ["https://play.google.com/store/apps/top?hl=en_GB"]),
    Topic(
      id: 9,
      text: "Praha",
      color: Colors.amberAccent,
      icon: Icon(Icons.location_city),
      questionIDs: range(202, 240),
    ),
    Topic(
      id: 10,
      text: "České památky",
      color: Colors.blue,
      icon: Icon(Icons.account_balance),
      questionIDs: range(241, 263, extra: {41, 46, 203, 204, 207, 221}),
    ),
    Topic(
        id: 11,
        text: "Čeští interpreti",
        color: Colors.purpleAccent,
        icon: Icon(Icons.music_note),
        questionIDs: range(266, 302),
        sources: ["http://www.zpev.hlasujpro.cz/zpevacky.html"]),
    Topic(
        id: 12,
        text: "Čeští herci a herečky",
        color: Colors.lime,
        icon: Icon(Icons.theaters),
        questionIDs: range(303, 342, extra: {292, 302}),
        sources: ["http://www.herec.hlasujpro.cz"]),
    Topic(
      id: 13,
      text: "Prezidenti ČR a ČS",
      color: Colors.yellow,
      icon: Icon(Icons.person),
      questionIDs: range(343, 353),
    ),
    Topic(
      id: 14,
      text: "Státy Evropy",
      color: Colors.blueAccent,
      icon: Icon(Icons.public),
      questionIDs: range(385, 434),
    ),
    Topic(
        id: 15,
        text: "National dishes",
        color: Colors.brown,
        icon: Icon(Icons.restaurant),
        questionIDs: range(437, 503),
        sources: ["https://en.wikipedia.org/wiki/National_dish#By_country"]),
    Topic(
      id: 16,
      text: "Nejlépe placení herci",
      color: Colors.amber,
      icon: Icon(Icons.star),
      questionIDs: range(504, 548),
    ),
    Topic(
      id: 17,
      text: "Nejlépe placení hudebníci",
      color: Colors.lightGreenAccent,
      icon: Icon(Icons.headset),
      questionIDs: range(550, 589),
    ),
    Topic(
      id: 18,
      text: "Nejkrásnější města ČR",
      color: Colors.pinkAccent,
      icon: Icon(Icons.location_city),
      questionIDs: range(591, 628, extra: {243}),
    ),
    Topic(
      id: 19,
      text: "Společenské hry",
      color: Colors.lightBlue,
      icon: Icon(Icons.extension),
      questionIDs: range(629, 649),
    ),
    Topic(
        id: 20,
        text: "Umělecké styly",
        color: Colors.tealAccent,
        icon: Icon(Icons.palette),
        questionIDs: range(650, 674),
        sources: ["https://cs.wikipedia.org/wiki/Kategorie:Um%C4%9Bleck%C3%A9_sm%C4%9Bry"]),
    Topic(
        id: 21,
        text: "Sporty",
        color: Colors.purple,
        icon: Icon(Icons.pool),
        questionIDs: range(675, 751, extra: {919, 955, 973, 980, 981, 983, 989, 991, 996, 1000, 1010, 1028}),
        sources: [
          "https://cs.wikipedia.org/wiki/Seznam_olympijsk%C3%BDch_sport%C5%AF",
          "https://cs.wikipedia.org/wiki/Abecedn%C3%AD_seznam_sportovn%C3%ADch_discipl%C3%ADn",
        ]),
    Topic(
      id: 22,
      text: "Zvířata",
      color: Colors.green,
      icon: Icon(Icons.pets),
      questionIDs: range(752, 900),
    ),
    Topic(
      id: 23,
      text: "Koníčky",
      color: Colors.pinkAccent,
      icon: Icon(Icons.favorite),
      questionIDs: range(904, 1028, extra: {
        675,
        679,
        680,
        682,
        684,
        688,
        690,
        691,
        692,
        695,
        696,
        701,
        703,
        705,
        707,
        708,
        710,
        719,
        724,
        726,
        727,
        728,
        730,
        731,
        732,
        733,
        734,
        735,
        736,
        737,
        742,
        743,
        744,
        745,
        746,
        748,
        749,
        750,
      }),
      sources: ["https://en.wikipedia.org/wiki/List_of_hobbies"],
    ),
    Topic(
        id: 24,
        text: "Videohry",
        color: Colors.yellow,
        icon: Icon(Icons.games),
        questionIDs: range(1029, 1108),
        sources: [
          "https://en.wikipedia.org/wiki/List_of_best-selling_video_games",
          "https://en.wikipedia.org/wiki/Arcade_game#List_of_highest-grossing_games",
          "https://en.wikipedia.org/wiki/List_of_video_games_by_monthly_active_player_count",
          "https://en.wikipedia.org/wiki/List_of_video_games_considered_the_best",
          "https://en.wikipedia.org/wiki/List_of_best-selling_game_consoles",
        ]),
    Topic(
        id: 25,
        text: "Jídlo a pití",
        color: Colors.brown,
        icon: Icon(Icons.restaurant_menu),
        questionIDs: range(1116, 1239),
        sources: [
          "https://en.wikipedia.org/wiki/Lists_of_foods",
          "https://en.wikipedia.org/wiki/Category:Appetizers",
          "https://en.wikipedia.org/wiki/Category:Condiments",
          "https://en.wikipedia.org/wiki/Category:Confectionery",
          "https://en.wikipedia.org/wiki/Category:Fast_food",
          "https://en.wikipedia.org/wiki/List_of_desserts",
          "https://en.wikipedia.org/wiki/List_of_street_foods",
          "https://en.wikipedia.org/wiki/Specialty_food",
          "https://en.wikipedia.org/wiki/List_of_drinks",
          "https://en.wikipedia.org/wiki/List_of_cocktails",
          "https://en.wikipedia.org/wiki/List_of_alcoholic_drinks",
        ]),
    Topic(
        id: 26,
        text: "Povolání",
        color: Colors.limeAccent,
        icon: Icon(Icons.work),
        questionIDs: range(1240, 1433),
        sources: ["https://www.prace.cz/encyklopedie-profesi/"]),
    Topic(
        id: 27,
        text: "Povinná četba",
        color: Colors.lightBlueAccent,
        icon: Icon(Icons.import_contacts),
        questionIDs: range(1434, 1567),
        sources: [
          "https://www.gvm.cz/images/stories/o-studiu/komise/cj/seznam-cetby-maturita-2019-2020.pdf",
        ]),
    Topic(
        id: 28,
        text: "Hudební nástroje",
        color: Colors.redAccent,
        icon: Icon(Icons.audiotrack),
        questionIDs: range(1568, 1612),
        sources: ["https://cs.wikipedia.org/wiki/Seznam_hudebn%C3%ADch_n%C3%A1stroj%C5%AF"]),
    Topic(
        id: 29,
        text: "Nejprodávanější singly",
        color: Colors.amber,
        icon: Icon(Icons.trending_up),
        questionIDs: range(1613, 1692),
        sources: ["https://en.wikipedia.org/wiki/List_of_best-selling_singles"]),
    Topic(
        id: 30,
        text: "České firmy",
        color: Colors.lime,
        icon: Icon(Icons.business_center),
        questionIDs: range(1693, 1742),
        sources: [
          "https://www.czechtop100.cz/cs/projekty/zebricky/100-nejvyznamnejsich",
          "https://cs.wikipedia.org/wiki/Kategorie:%C4%8Cesk%C3%A9_firmy",
        ]),
  ],
  key: (topic) => topic.id,
  value: (topic) => topic,
);
